File: kythe/go/indexer/testdata/basic/writes.go
// Package writes tests ref/writes edges
package writes

import "fmt"

type S struct {
	//- @F defines/binding F
	F int

	//- @Nested defines/binding Nested
	Nested *S
}

func f() {
	//- @x defines/binding X
	var x int

	//- @x ref/writes X
	//- @y defines/binding Y
	x, y := 1, 2

	//- @x ref X
	//- @x ref/writes X
	x += 42

	//- @y ref Y
	//- @y ref/writes Y
	y++

	//- @#0i defines/binding I
	//- @#1i ref I
	//- @#2i ref/writes I
	//- @#3i ref I
	for i := 0; i < 10; i = i + 1 {
	}

	//- @z defines/binding Z
	//- @Nested ref/writes Nested
	//- !{ @Nested ref Nested }
	z := S{Nested: &S{}}
	//- @z ref Z
	//- @F ref/writes F
	z.F = 42
	//- @z ref Z
	//- @Nested ref Nested
	//- @F ref/writes F
	z.Nested.F = 52

	//- @x ref X
	//- @y ref Y
	fmt.Println(x, y)
}

func g() any {
	//- @val defines/binding Val
	var val int

	return map[int]bool{
		//- @val ref Val
		//- !{ @val ref/writes Val }
		val: true,
	}
}

File: kythe/typescript/testdata/writes.ts
class WritesClass {
  //- @member defines/binding Member
  member: number;

  //- @sub defines/binding MemberSub
  sub: WritesClass;

  constructor() {
    //- !{ @member ref Member }
    //- @member ref/writes Member
    this.member = 1;
    //- !{ @sub ref MemberSub }
    //- @sub ref/writes MemberSub
    this.sub = new WritesClass();
  }
}

//
// Test Assignments
//

//- @w defines/binding W
let w = 1;

//- !{ @w ref W }
//- @w ref/writes W
w = 2;

//- !{ @#0w ref W }
//- @#0w ref/writes W
//- @#1w ref W
//- !{ @#1w ref/writes W }
w = w + 2;

//- @x defines/binding X
let x = 2;

//- !{ @x ref X }
//- @x ref/writes X
//- @w ref W
//- @w ref/writes W
x = w = 3;

//- !{ @#0x ref X }
//- @#0x ref/writes X
//- @#1x ref X
//- @#1x ref/writes X
x = x++ + 1;

//- @y defines/binding Y
let y = new WritesClass();

//- @y ref Y
//- !{ @y ref/writes Y }
//- !{ @member ref Member }
//- @member ref/writes Member
y.member = 1;

//- !{ @#0member ref Member }
//- @#0member ref/writes Member
//- @#1member ref Member
//- !{ @#1member ref/writes Member }
y.member = y.member + 1;

//
// Test pre/postfix operations
//

//- @w ref W
//- @w ref/writes W
w++;
//- @w ref W
//- @w ref/writes W
w--;
//- @w ref W
//- @w ref/writes W
++w;
//- @w ref W
//- @w ref/writes W
--w;

//- @member ref Member
//- @member ref/writes Member
y.member++;
//- @member ref Member
//- @member ref/writes Member
y.member--;
//- @member ref Member
//- @member ref/writes Member
++y.member;
//- @member ref Member
//- @member ref/writes Member
--y.member;

//
// Test compound assignment token
//

//- @w ref W
//- @w ref/writes W
w += 1;
//- @w ref W
//- @w ref/writes W
w -= 1;
//- @w ref W
//- @w ref/writes W
w *= 1;
//- @w ref W
//- @w ref/writes W
w **= 1;
//- @w ref W
//- @w ref/writes W
w /= 1;
//- @w ref W
//- @w ref/writes W
w %= 1;
//- @w ref W
//- @w ref/writes W
w <<= 1;
//- @w ref W
//- @w ref/writes W
w >>= 1;
//- @w ref W
//- @w ref/writes W
w >>>= 1;
//- @w ref W
//- @w ref/writes W
w &= 1;
//- @w ref W
//- @w ref/writes W
w |= 1;
//- @w ref W
//- @w ref/writes W
w ^= 1;

//
// Test chained property access
//

//- @sub ref MemberSub
//- !{ @sub ref/writes MemberSub }
//- !{ @member ref Member }
//- @member ref/writes Member
y.sub.member = 1;

//
// Some more complicated cases
//

//- @z defines/binding Z
let z = 1;

//- !{ @z ref Z }
//- @z ref/writes Z
//- @x ref X
//- @x ref/writes X
//- @#0w ref W
//- @#0w ref/writes W
//- @#1w ref W
//- !{ @#1w ref/writes W }
z = x = w++ + w + 1;

//- @#0z ref Z
//- @#0z ref/writes Z
//- @#1z ref Z
//- !{ @#1z ref/writes Z }
z += z;
